slam/src/slam/fast_slam2/FastSLAM2Particle.py

```diff
@@ -0,0 +1,246 @@
+    Currently assumes known correspondences.
+    '''
+    self.weight = 1.0
+
+    # If there are no measurements, sample the pose according to the default pose covariance and the pose mean computed by the motion model
+    if len(meas_ls) == 0:
+      self.pose = self._robot_physics.compute_sample_pose(self.pose, self.pose_cov)
+      return
+    
+    # Compute the particle's weight
+    for meas in meas_ls:
+      observed_landmark = self.landmarks[meas.correspondence]
+      est_meas_data = self._robot_physics.compute_meas_model(self.pose, observed_landmark.mean)
+      self.weight *= scipy.stats.multivariate_normal.pdf(est_meas_data, mean=meas.data, cov=meas.cov)
+
+    # Bound the weight above a tiny number
+    self.weight = max(self.weight, 1e-50)
+
+  def _update_dependencies_on_landmark(self, landmark: _EKF,  pose_mean, pose_cov, meas):
+    est_meas_data = self._robot_physics.compute_meas_model(pose_mean, landmark.mean)
+    landmark.meas_jac_pose, landmark.meas_jac_land = self._robot_physics.compute_meas_jacobians(pose_mean, landmark.mean)
+    landmark.innovation = meas.data - est_meas_data
+
+    landmark.inv_pose_cov = np.linalg.inv(pose_cov) 
+    landmark.Q = meas.cov + landmark.meas_jac_land @ landmark.cov @ landmark.meas_jac_land.T
+    landmark.Q_inv = np.linalg.inv(landmark.Q)
+    
+  def _update_pose_distribution(self, landmark, pose_mean, pose_cov):
+    landmark.pose_cov = np.linalg.inv(landmark.meas_jac_pose.T @ landmark.Q_inv @ landmark.meas_jac_pose + landmark.inv_pose_cov)
+    landmark.pose_mean = pose_cov @ landmark.meas_jac_pose.T @ landmark.Q_inv @ landmark.innovation + pose_mean
+    landmark.sampled_pose = self._random.multivariate_normal(pose_mean, pose_cov)
+
+  def _update_landmark_association_prob(self, meas, landmark):
+    improved_meas_data = self._robot_physics.compute_meas_model(landmark.sampled_pose, landmark.mean) #range_improved, bearing_improved
+    improved_innovation = meas.data - improved_meas_data
+    
+    exponent = -0.5 * (improved_innovation).T @ landmark.Q_inv @ improved_innovation
+    inv_sqrt_of_two_pi_det_Q = (np.linalg.det(2 * np.pi * landmark.Q)) ** -0.5
+    landmark.association_prob = inv_sqrt_of_two_pi_det_Q * np.exp(exponent)
+
+  def _update_observed_landmark(self,landmark, pose_cov, meas):
+    landmark.exist_log += self._landmark_constants.exist_log_inc
+    # compute joint distribution (Kalman gain)
+    prev_landmark_cov = landmark.cov
+    K =  prev_landmark_cov @ landmark.meas_jac_land.T @ landmark.Q_inv 
+    I = np.eye(landmark.mean.shape[0])
+    # update landmark
+    landmark.mean = landmark.mean + K @ landmark.innovation
+    
+    landmark.cov = ( I - K @ landmark.meas_jac_land) @ prev_landmark_cov
+    # Compute particle weight
+    L = landmark.meas_jac_pose @ pose_cov @ landmark.meas_jac_pose.T \
+         + landmark.meas_jac_land @ prev_landmark_cov @ landmark.meas_jac_land.T \
+         + meas.cov
+    L_inv = np.linalg.inv(L)
```

Comment:
Same as i mentioned before, I usually find just calling `inv` without checking if it's possible to be a bit scary, but it should __usually__ be fine.

Reply:
We will check for singular matrix, and register a warming and use pseudo-inverse if the matrix is singular.